Fix: Issue #16 (description: Restrict creating mines more than total cells)

b0r1ngx opened the pull request:
Solution of a problem raised at Issue https://github.com/arkivanov/Minesweeper/issues/16

Closes #16.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit


- **Bug Fixes**
	- Ensured the maximum number of mines in Minesweeper does not exceed the total number of cells.
	- Improved validation for width, height, and maximum mines to fall within specific ranges.
	- Introduced `coerceIn` to ensure proper range for width, height, and max mines.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

b0r1ngx committed:
Fix: Setting GameSettings to impossible states doesn't provide good user interface feedback #16

b0r1ngx commented:
How do you like this solution of Issue #16 ?

Also I'm not so much work with flow like that and I don't know or I just can't to assign myself to issue, to show that issue have this pull request as problem solution, I think I just need to read about it somewhere ;)

arkivanov commented:
@coderabbitai review

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/settings/DefaultEditSettingsComponent.kt:
```diff
@@ -44,7 +45,7 @@ internal class DefaultEditSettingsComponent(
                 width = width,
                 height = height,
-                maxMines = maxMines,
+                maxMines = min(maxMines, width * height),
```
It would be nice to also validate width and height to be minimum 2. And a maximum like 100? We can also use coerceAtLeast , coerceAtMost and coerceIn - a Kotlin idiomatic way.

However, does it make sense to have maxMines equal to width * height? Such a game is 100% fail from the first click. I suggest to validate as follows.

`maxMines.coerceIn(1, width * height - 1)`

What do you think?

b0r1ngx commented:
Waow, new instruments, awesome I'm checked out today, not get any notifies by email:\

b0r1ngx commented:
You have a nice conversation here with some of LLM 😅

> However, does it make sense to have maxMines equal to width * height? Such a game is 100% fail from the first click. I suggest to validate as follows.

About this idea: yea, I'm also tried use width * height - 1, later user's can compete for who is fastest make first click on the field, it's really sounds funny, for someone

P.S. game not fails but became to win state, maybe if we save: width * height, user's can compete by who have faster chip, system that is run this project.

b0r1ngx commented:
> It would be nice to also validate width and height to be minimum 2. And a maximum like 100? We can also use coerceAtLeast , coerceAtMost and coerceIn - a Kotlin idiomatic way.

Yes, we can do that, lets me make a commit for this change.

I have a question about this change, should we somehow show information about limitations to user that they are allowed to play only at 10 < width/height <= 100 setting?

For example if they entered 5, we highlight textfiled with red and show hint with allowed settings?

arkivanov commented:
> I have a question about this change, should we somehow show information about limitations to user that they are allowed to play only at 10 < width/height <= 100 setting?

> For example if they entered 5, we highlight textfiled with red and show hint with allowed settings?

Let's keep it simple for now. From my point of view it makes sense to just coerce width between 2 and 100 and height between 2 and 50. What do you think?

b0r1ngx commented:
Also entering negative number leads to: 
<img width="586" alt="image" src="https://github.com/arkivanov/Minesweeper/assets/55315908/356b980f-9fa4-49af-9255-f827c0bbabaa">

I'm fix it in this PR.

b0r1ngx commented:
> What do you think?

Yes, lets do this!

arkivanov commented:
> Also entering negative number leads to: <img alt="image" width="586" src="https://private-user-images.githubusercontent.com/55315908/314255236-356b980f-9fa4-49af-9255-f827c0bbabaa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTA4ODg3MzksIm5iZiI6MTcxMDg4ODQzOSwicGF0aCI6Ii81NTMxNTkwOC8zMTQyNTUyMzYtMzU2Yjk4MGYtOWZhNC00OWFmLTkyNTUtZjgyN2MwYmJhYmFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAzMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMzE5VDIyNDcxOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTdiZWIwODMxN2I4NmJlYmZmMjE4YzkxYTBhMTc0NTk3ZTk0NTZjMzE3MWE1NTRkODlkNGU5NGYyN2I1MDJkOGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.lO-taxTarp8CHbTSbR5c8vJQcqzBbXprDJTKptOTr6k">
> 
> I'm fix it in this PR.

Yep, that's why I suggest to

```kotlin
maxMines.coerceIn(1, width * height - 1)
```

b0r1ngx committed:
fixed apply settings

b0r1ngx commented:
> maxMines.coerceIn(1, width * height - 1)

We must use something like I did in my last commit, because if we just coerce width / height values and not saving it for later usage for maxMines (like in replied message), we can get error, because width/height isn't fixed (coerced).

b0r1ngx committed:
Refactoring (Magic Numbers)

b0r1ngx commented:
@coderabbitai I pushed a fix in commit [d762256](https://github.com/arkivanov/Minesweeper/pull/17/commits/d7622568aeed2ef8dc5487b2f7449fcf3db917d8)

b0r1ngx commented:
> Tip
> 
> For best results, initiate chat on the files or code changes.
> 
> @b0r1ngx, I've reviewed the changes made in commit [d762256](https://github.com/arkivanov/Minesweeper/commit/d7622568aeed2ef8dc5487b2f7449fcf3db917d8) to the `DefaultEditSettingsComponent.kt` file. The implementation effectively addresses the issues discussed:
> 
> 1. **Limiting Mines**: The `maxMines` value is coerced to be within the range of 1 to one less than the product of `width` and `height`, ensuring that the game cannot be set up with more mines than there are cells available, minus one. This aligns with the suggestion to prevent a game setup where the first click would result in a loss.
> 2. **Handling Negative Numbers for Width and Height**: Both `width` and `height` are coerced to be within reasonable limits (2 to 100 for width and 2 to 50 for height), which implicitly handles negative inputs by setting them to the minimum allowed value.
> 
> This solution appears to be a robust way to ensure game settings are within logical and playable limits, enhancing the overall user experience. Great job on implementing these changes!
> 
> Analysis chain
> Scripts executed

@coderabbitai can you give any advice how to increase code maintainability in this commit?

b0r1ngx commented:
@coderabbitai I pushed a fix in commit [069a93b](https://github.com/arkivanov/Minesweeper/pull/17/commits/069a93b2d30a9c2dfd20d1dad17ed2ca3bd95769)

What is the purpose of this commit? 
What code metrics is increasing?

Give answer for each question.

b0r1ngx commented:
>  For example, coerceIn(min = 1, max = width * height - 1) makes it immediately clear what the 1 and width * height - 1 values represent.

@arkivanov What do you think about this LLM advice? Should it be done or it's not improve code maintainability? Maybe it entails some disadvantages?

We can impove it with something like:
`private const val MIN_MINES = 1`

arkivanov commented on composeApp/src/commonMain/kotlin/com/arkivanov/minesweeper/settings/DefaultEditSettingsComponent.kt:
```diff
@@ -40,13 +44,11 @@ internal class DefaultEditSettingsComponent(
-        )
+        val finalWidth = width.coerceIn(MIN_SIZE..MAX_WIDTH)
+        val finalHeight = height.coerceIn(MIN_SIZE..MAX_HEIGHT)
+        val finalMines = maxMines.coerceIn(1 until finalWidth * finalHeight - 1)
```
I think `until` is exclusive on the right side. We could write `coerseIn(1, finalWidth * finalHeight - 1)`, or `coerseIn(1 until finalWidth * finalHeight)`.

arkivanov commented:
> What do you think about this LLM advice? Should it be done or it's not improve code maintainability? Maybe it entails some disadvantages?

I think the current code is good enough, no need to use named arguments in that particular case. Personally, I wouldn't extract constants either - `2..100` looks more readable to me. But it's just my personal opinion.

b0r1ngx commented:
Thanks for feedback! I agree with you!

b0r1ngx committed:
reverse refactoring of magic numbers - in case that constants in code is locally understandable what it is used about, without any additional explanations about it (like MIN_SIZE, etc...)

arkivanov commented:
Thank you!
